### DISCUSSION TITLE
Port of carbon components

### DISCUSSION THREAD

**Discussion by mosheduminer** [carbon-design-system/carbon#8759] [Help] [answered]
[I originally posted this in the discord a couple days ago, but I did get a response. I got the feeling that the discord is not really monitored, so I am asking again here]

Background: I wish to port Carbon Components to a framework called Solid (https://github.com/solidjs/solid). Since Solid has a lot of APIs that are fairly similar to React's, I decided to simply copy components from carbon-components-react, and modify them as necessary. All good so far, ported just the Accordion component(s).

Now the question: Apache License 2.0 (the license used by carbon) says (4b):
> You must cause any modified files to carry prominent notices stating that You changed the files;

I think this means that I need to put a notice in every source file in my project (since they are almost all going to be modifications of the corresponding React files)? Is this correct?

While looking into this, I noticed that carbon-design-system hosts a number of community ports (?) of carbon components. What is the process for being hosted as such? Could my port be hosted under carbon-design-system? (if yes, I assume that would affect licensing concerns).

**Comment by jnm2377** [comment, marked as answer]
Hey @mosheduminer here's a similar discussion that talks about the limitations of the Apache license. https://github.com/carbon-design-system/carbon/discussions/7657

Basically you just have to make it clear that you are not Carbon in every file. If you notice, we do this with out files. All of our files have a copyright trademark at the top. So something like that that states you've modified it.

**Reply by jnm2377** [reply]
@mosheduminer we'd also be happy to set something up if you're still interested in contributing a community maintained version of Carbon for Solid.js.

**Reply by mosheduminer** [reply]
I would indeed be interested, but I would need to find additional maintainers first. I ported perhaps 20% of the components (no tests or docs) before my pace slowed, which showed me that I can't always be relied upon to manage this alone.

**Reply by mosheduminer** [reply]
Hey, I know it's been a long time, but I ported probably over half the components (including some complex ones like UIShell and DataTable, currently all in a private repository), mostly by myself. I think the only way the port will be able to really go forward is as a carbon community project, which will hopefully draw more maintainers as interest grows. What do you think? Can we set something up?

**Reply by tay1orjones** [reply]
@mosheduminer Thanks for helping out! Could you make the repo public so we can take a look? I'm not opposed to moving the repo into the organization if there's a decent amount of fidelity and commitment to support.

**Reply by mosheduminer** [reply]
Sure! I'll do that later today, as well as deploy a basic example site.
My primary concern has been complying with licensing when I make it public. For now I intend to simply copy the "upstream" apache license and refer to it, rather than including a license/reference in each file.

Thank you.

**Reply by tay1orjones** [reply]
> as well as deploy a basic example site

@mosheduminer So far the bulk of our framework options have support from Storybook. Would that be an option for this?

**Reply by mosheduminer** [reply]
So far I haven't tried the storybook option, as storybook + solid has been difficult to get working in the past (although it is possible), and I have been prioritizing working code rather than documentation. That said, I intend to either get vitebook support for solidjs (which can serve as a storybook alternative) or setup a storybook.

The example site I speak of is really barebones, and already exists with a subset of the ported components (it is just is not deployed yet). I don't intend to keep it long term, but I can spin it up in a few minutes.

**Reply by mosheduminer** [reply]
@tay1orjones here's the repo for when you have the time to look at it: https://github.com/mosheduminer/carbon-components-solid. There's a site I spun up quickly linked in the description. The code is usually pretty close to the react version.

**Comment by davedbase** [comment]
@jnm2377 hi! I was wondering if we can get this community version of SolidJS going? It sounds like @mosheduminer did a lot of work porting it and I'd love to see how to take advantage of the Carbon/Solid combo.

Please support community contributions :-)

**Reply by tay1orjones** [reply]
@davedbase Definitely! This isn't blocked by the team in any way right now, @mosheduminer's repo is private.

**Comment by tay1orjones** [comment]
Is this related to `carbon-components-solid` at all? https://gitlab.com/mneumann_ntecs/carbon-components-solid

Are there any other community projects focused on Solid?

**Reply by mosheduminer** [reply]
No, this is not related to that repository, my work is completely independent. Unfortunately they already have the NPM name for `carbon-components-solid`. I don't believe there is anyone else who has worked on a solid carbon port (if they have, they haven't made it public or mentioned it on the official discord).
